### PR TITLE
fix(generic-metrics): Add SnubaHTTPBackend to Metrics Interface for testing

### PR DESCRIPTION
**Pull request by ayirr7**
Follow-up work an issue where we were recreating the producer instance of this interface upon each emitted metric. This PR uses the LazyServiceWrapper in order to abstract away from the user which backend is being used, and also enable the usage where the backend is instantiated only once per "metric emitting" session. Tests have been modified as well to reflect this.

Update: Because Kafka is not running in CI/dev, we need to create another backend of the abstract `GenericMetricsBackend`. This would be a backend that writes directly to Clickhouse using the HTTP endpoint exposed in Snuba

This work will be done in a couple of iterations, the first one lives in this PR. Though the goal is for this SnubaBackend to ultimately replace `store_metric` in the Metrics Layer, in this iteration, we will use `store_metric` in the code in order to unblock any product teams who want to emit metrics. 

This PR:
- Adds this Snuba HTTP backend, making it the default as well via `server.py`
- Adds a test for this backend. This test is adapted from similar tests in the Metrics Layer
- Fixes parts of `store_metric` so that it's compatible with new `UseCaseID`s

### CODE REVIEW

File: tests/sentry/sentry_metrics/test_kafka.py [hunk 1]
```diff
@@ -18,19 +19,22 @@
 tags = {"a": "b"}
 
 
+metrics_backend = client.backend
+# For testing, we are calling close() here because we
+# are swapping out the KafkaProducer
+# with a LocalProducer, but regardless,
+# close() must always be called in order to close
+# the backend's KafkaProducer
+metrics_backend.close()
```
Comment (untitaker) [inline tests/sentry/sentry_metrics/test_kafka.py:28]: please do this in a fixture or setup/teardown callback, not on import

**Review by untitaker — APPROVED** [review]

File: src/sentry/sentry_metrics/client/__init__.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+from sentry.utils.services import LazyServiceWrapper
+
+from .base import GenericMetricsBackend
+
+backend = LazyServiceWrapper(
+    GenericMetricsBackend,
+    settings.SENTRY_METRICS_INTERFACE_BACKEND,
+    settings.SENTRY_METRICS_INTERFACE_BACKEND_OPTIONS,
+)
+backend.expose(locals())
```
Comment (untitaker) [inline src/sentry/sentry_metrics/client/__init__.py:12]: i'd like to stop using `expose` as it is not typesafe. can we make people use methods on `backend` directly?

**Comment by ayirr7** [comment]
The changes in the post_process are to actually simulate the usage of the metrics backend and see if it works as expected. Will remove that piece of code once verified it works.

File: src/sentry/tasks/post_process.py [hunk 3]
```diff
@@ -571,6 +588,7 @@ def get_event_raise_exception() -> Event:
         update_event_groups(event, group_states)
         bind_organization_context(event.project.organization)
         _capture_event_stats(event)
+        _update_escalating_metrics(event)
```
Comment (lynnagara) [inline src/sentry/tasks/post_process.py:591]: What's this change about? It seems unrelated to changing the metrics interface?

File: src/sentry/sentry_metrics/client/base.py [hunk 4]
```diff
@@ -64,3 +61,10 @@ def distribution(
         has been registered in the UseCaseID enum.
         """
         raise NotImplementedError()
+
+    def close(self) -> None:
+        """
+        Calling this is required once we are done emitting metrics
```
Comment (lynnagara) [inline src/sentry/sentry_metrics/client/base.py:67]: This states it is required, but we are not actually calling it anywhere.

File: src/sentry/sentry_metrics/client/__init__.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+from .base import GenericMetricsBackend
+
+generic_metrics_backend = LazyServiceWrapper(
+    GenericMetricsBackend,
+    settings.SENTRY_METRICS_INTERFACE_BACKEND,
+    settings.SENTRY_METRICS_INTERFACE_BACKEND_OPTIONS,
+)
+
+__all__ = ["generic_metrics_backend"]
```
Comment (lynnagara) [inline src/sentry/sentry_metrics/client/__init__.py:13]: nit: i'd just call this "backend" since it's already inside the metrics module?

File: src/sentry/tasks/post_process.py (hunk 3 above)
Comment (ayirr7) [reply src/sentry/tasks/post_process.py:591]: I was just copying the code that the product team had written in order to temporarily see how it behaves with the metrics backend. would remove it once i verify its fine.

**Comment by lynnagara** [comment]
Unfortuantely we need to work around the fact that Kafka is optional in dev and some tests. In these environments Snuba exposes a test API that can be used as an alternative to producing to Kafka, which we typically use instead: https://github.com/getsentry/snuba/blob/745d81480f10fcddc89839a27ed0c3e131fde9f7/snuba/web/views.py#L555-L595

If you take a look at the `eventstream` module, there are two backends - `snuba` and `kafka` where the Snuba one uses the HTTP endpoint and is only for testing. I'd recommend taking the same approach and making `snuba` default just like it is for eventstream.

Also, in the this `kafka` backend, please lazily instantiate the producer on the first call to counter()/set()/distributiones()/whatever, and use the same instance thereafter to avoid any unnecessary connections

**Comment by ayirr7** [comment]
> Also, in the this kafka backend, please lazily instantiate the producer on the first call to counter()/set()/distributiones()/whatever, and use the same instance thereafter to avoid any unnecessary connections

I think this is already being done based on the `LazyServiceWrapper` implementation. As soon as we access any attribute of the backend (such as the call `generic_metrics_backend.counter()`) the underlying instance (a `KafkaMetricsBackend`) gets initialized. And that point on, any attribute is accessed via that same instance

**Comment by onewland** [comment]
> Unfortuantely we need to work around the fact that Kafka is optional in dev and some tests. In these environments Snuba exposes a test API that can be used as an alternative to producing to Kafka, which we typically use instead: https://github.com/getsentry/snuba/blob/745d81480f10fcddc89839a27ed0c3e131fde9f7/snuba/web/views.py#L555-L595
> 
> If you take a look at the `eventstream` module, there are two backends - `snuba` and `kafka` where the Snuba one uses the HTTP endpoint and is only for testing. I'd recommend taking the same approach and making `snuba` default just like it is for eventstream.
> 

I actually think this is a much bigger change than you're imagining. This "backend" produces to the metrics indexer in Sentry, which does not have a swappable transport layer (and I'm not sure we should build one for it either) for forwarding data on to Snuba. If we want to support that in the indexer, I don't think you want to combine it with this change

**Comment by untitaker** [comment]
@onewland had a call with riya on friday and suggested that the "http backend" calls the indexer consumer's message processor to transform the payloads, or does something similar to `TestCase.store_metric` in that it calls the indexer.record function directly. I think having this backend use the consumer code would be really cool though because it allows us to get rid of the shoddy reimplementation that is `store_metric`.

File: src/sentry/sentry_metrics/client/snuba.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        produced to the broker yet.
+        """
+        dist_metric = {
+            "mapping_meta": self.fake_mapping_meta,
+            "metric_id": self.metric_id,
+            "org_id": org_id,
+            "project_id": project_id,
+            "value": value,
+            "timestamp": int(datetime.now().timestamp()),
+            "tags": tags,
```
[the remark this answers is not in the document]
Comment (ayirr7) [reply src/sentry/sentry_metrics/client/snuba.py:151]: yep, that initial commit was just to get a scaffolding ready to get an idea of what it would look like

**Comment by ayirr7** [comment]
Still need to write some unit tests that will send metrics to the Snuba endpoint

**Comment by ayirr7** [comment]
Commented out `test_kafka.py` for now because the default backend being used is the SnubaBackend, not the KafkaBackend.  Will see what we can do with that test.

File: tests/sentry/sentry_metrics/test_kafka.py [hunk 7]
```diff
@@ -1,153 +1,127 @@
...
-        "type": "d",
-    }
-
-    value = json.dumps(distribution_metric).encode("utf-8")
-
-    produced_message = broker_storage.consume(Partition(my_topic, 0), 0)
-    assert produced_message is not None
-    assert produced_message.payload.value == value
-    assert broker_storage.consume(Partition(my_topic, 0), 1) is None
+# from datetime import datetime
```
Comment (untitaker) [inline tests/sentry/sentry_metrics/test_kafka.py:1]: this entire file is commented out, i suppose that's not intended?

File: src/sentry/sentry_metrics/client/base.py (hunk 4 above)
Comment (untitaker) [reply src/sentry/sentry_metrics/client/base.py:67]: I believe we use this in tests now, but if we don't we might want to remove it

File: src/sentry/sentry_metrics/client/snuba.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+from sentry import quotas
+from sentry.sentry_metrics.client.base import GenericMetricsBackend
+from sentry.sentry_metrics.configuration import IndexerStorage, UseCaseKey, get_ingest_config
+from sentry.sentry_metrics.consumers.indexer.processing import MessageProcessor
+from sentry.sentry_metrics.consumers.indexer.routing_producer import RoutingPayload
+from sentry.sentry_metrics.use_case_id_registry import UseCaseID
+from sentry.utils import json, snuba
+
+# ingest_codec: sentry_kafka_schemas.codecs.Codec[Any] = sentry_kafka_schemas.get_codec(
```
Comment (untitaker) [inline src/sentry/sentry_metrics/client/snuba.py:19]: dead code to delete?

File: src/sentry/sentry_metrics/client/snuba.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+            "tags": tags,
+            "retention_days": get_retention_from_org_id(org_id),
+            "type": "d",
+        }
+
+        self.__build_and_send_request(dist_metric)
+
+    def __build_payload(self, metric) -> Sequence[Any]:
+        # build message batch for the MessageProcessor
+        message_batch = [
```
Comment (untitaker) [inline src/sentry/sentry_metrics/client/snuba.py:160]: *nit:* constructing kafka-specific things like this is pretty gross fwiw but i think it is still much better than duplicating biz logic

ideally we'd clean up the interface for MessageProcessor so that this is not necessary (def not as part of this PR)

File: src/sentry/tasks/post_process.py (hunk 3 above)
Comment (untitaker) [reply src/sentry/tasks/post_process.py:591]: just to highlight this again, please revert these changes

File: tests/sentry/sentry_metrics/test_snuba.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+            "organization_id": self.organization.id,
+            "project_id": self.projects,
+            "start": datetime.utcnow() - timedelta(days=1),
+            "end": datetime.utcnow() + timedelta(days=1),
+        }
+
+
+class MetricsInterfaceTest(MetricsInterfaceTestCase):
+    def test(self):
+        generic_metrics_backend.set(
```
Comment (untitaker) [inline tests/sentry/sentry_metrics/test_snuba.py:33]: I would recommend reusing the testcases between kafka and snuba backend. typically we do this using `pytest.mark.parametrize` or by creating a mixin class of testcases that we reuse in both testsuites

**Review by untitaker — COMMENTED** [review]
non-final pass of review, mostly cleanup

File: tests/sentry/sentry_metrics/test_snuba.py (hunk 10 above)
Comment (ayirr7) [reply tests/sentry/sentry_metrics/test_snuba.py]: done

**Comment by ayirr7** [comment]
Snuba backend has been updated to use `store_metric` for now. The corresponding test also uses it + Metric Layer query API to verify that metrics are produced to the right table successfully. The next iteration would be to actually build out the steps for producing to the Snuba endpoint directly instead of using `store_metric`.

File: src/sentry/sentry_metrics/client/snuba.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+from typing import Dict, Optional, Sequence, Union
+
+from django.core.cache import cache
+
+from sentry import quotas
+from sentry.sentry_metrics.client.base import GenericMetricsBackend
+from sentry.sentry_metrics.configuration import IndexerStorage, UseCaseKey, get_ingest_config
+from sentry.sentry_metrics.consumers.indexer.processing import MessageProcessor
+from sentry.sentry_metrics.use_case_id_registry import UseCaseID
+from sentry.testutils.cases import BaseMetricsTestCase as metrics_test_base
```
Comment (untitaker) [inline src/sentry/sentry_metrics/client/snuba.py:13]: nit: please don't rename this class to snake_case on import

File: src/sentry/sentry_metrics/client/snuba.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    """
+    This backend is meant for use in dev/testing
+    environments. It allows for producing metrics
+    to a Snuba HTTP endpoint, which will trigger
+    processing and ultimately, insertion, of the
+    metric into Clickhouse.
+    """
+
+    def __init__(self) -> None:
+        self._message_processor = MessageProcessor(
```
Comment (untitaker) [inline src/sentry/sentry_metrics/client/snuba.py:47]: this seems unused

**Review by untitaker — APPROVED** [review]

File: src/sentry/sentry_metrics/client/snuba.py (hunk 11 above)
Comment (ayirr7) [reply src/sentry/sentry_metrics/client/snuba.py]: done

File: src/sentry/sentry_metrics/client/snuba.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+
+from datetime import datetime
+from typing import Dict, Optional, Sequence, Union
+
+from django.core.cache import cache
+
+from sentry import quotas
+from sentry.sentry_metrics.client.base import GenericMetricsBackend
+from sentry.sentry_metrics.use_case_id_registry import UseCaseID
+from sentry.testutils.cases import BaseMetricsTestCase
```
Comment (onewland) [inline src/sentry/sentry_metrics/client/snuba.py:11]: Do we import from testutils in application code elsewhere? It has a bit of a code smell to me
